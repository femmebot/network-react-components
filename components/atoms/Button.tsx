import MuiButton from '@material-ui/core/Button/Button'
import * as React from 'react'
import styled from 'styled-components'
import {colors} from '~shared/styles/index'
import {pxToRem} from '~shared/styles/utils'

interface BaseProps {
  disabled?: boolean
  variant?: Variant
  wide?: boolean
}
interface LinkButtonProps extends BaseProps {
  href: string
  target?: '_blank'
}

export interface ButtonProps extends BaseProps {
  onClick?: (event: any) => void
  type?: 'button' | 'submit'
  name?: string
}

export type Props = ButtonProps | LinkButtonProps

type Variant = 'outline' | 'commit'

const outlineOverrides = () => `
  border: solid ${pxToRem(2)} currentColor;
  border-radius: ${pxToRem(2)};
`

const commitOverrides = () => `
  background: ${colors.ctaPrimary};
  &:hover {
    background: ${colors.ctaSecondary};
  }
}
`

const variantOverrides = (variant?: Variant) => {
  if (!variant) {
    return
  }

  if (variant === 'outline') {
    return outlineOverrides()
  }

  if (variant === 'commit') {
    return commitOverrides()
  }

  return
}

const StyledButton = styled(
  ({variant: _variant, wide: _wide, ...props}: Props) => (
    <MuiButton variant="flat" {...props} />
  )
)`
  && {
    display: inline-flex;
    align-items: center;
    color: ${(props: Props) => (props.disabled ? 'gray' : 'black')};
    padding: ${pxToRem(8)};
    min-width: ${(props: Props) => (props.wide ? pxToRem(220) : undefined)};
    svg {
      margin-right: ${pxToRem(8)};
    }
    ${(props: Props) => variantOverrides(props.variant)};
  }
`

const Button: React.SFC<Props> = ({children, variant, ...props}) => (
  <StyledButton variant={variant} {...props}>
    {children}
  </StyledButton>
)

export default Button
