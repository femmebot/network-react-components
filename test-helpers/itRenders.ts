import 'jest-styled-components';
import forEach from 'lodash/forEach';
import renderer from 'react-test-renderer';

export const expectSnapshotMatches = (node: JSX.Element) => {
  const tree = renderer.create(node).toJSON();
  expect(tree).toMatchSnapshot();
};

const makeSpec = (title: string, example: Example) => {
  it(title, () => {
    const tree = renderer.create(example()).toJSON();
    expect(tree).toMatchSnapshot();
  });
};

type Example = () => JSX.Element;
type Examples = { [when: string]: Example } | Example;

const itRenders = (examples: Examples) => {
  if (typeof examples === 'function') {
    makeSpec('renders', examples);
  } else {
    forEach(examples, (example, when) => makeSpec(`renders ${when}`, example));
  }
};

export default itRenders;
