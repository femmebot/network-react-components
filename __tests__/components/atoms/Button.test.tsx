import React from 'react';
import itRenders from '~shared/test-helpers/itRenders';

import Button from '~shared/components/atoms/Button';

itRenders({
  default: () => <Button onClick={jest.fn()}>Content</Button>,
  disabled: () => (
    <Button disabled onClick={jest.fn()}>
      Disabled
    </Button>
  ),
  'as a link': () => (
    <Button disabled href="#">
      Link
    </Button>
  ),
  'submit type': () => <Button type="submit">Submit</Button>,
  wide: () => (
    <Button wide href="#">
      Link
    </Button>
  ),
  'outline variant': () => (
    <Button variant="outline" href="#">
      Link
    </Button>
  ),
  'commit variant': () => <Button variant="commit">Commit</Button>,
});
