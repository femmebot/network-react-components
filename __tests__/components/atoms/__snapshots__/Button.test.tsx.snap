// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as a link 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: gray;
  padding: 0.5rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<a
  className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 MuiButton-disabled-21 c0"
  href="#"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="-1"
>
  <span
    className="MuiButton-label-2"
  >
    Link
  </span>
</a>
`;

exports[`renders commit variant 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: black;
  padding: 0.5rem;
  background: #EED950;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

.c0.c0:hover {
  background: #F6ECA8;
}

<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 c0"
  disabled={false}
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    Commit
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;

exports[`renders default 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: black;
  padding: 0.5rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 c0"
  disabled={false}
  onBlur={[Function]}
  onClick={[MockFunction]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    Content
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;

exports[`renders disabled 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: gray;
  padding: 0.5rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<button
  className="MuiButtonBase-root-27 MuiButtonBase-disabled-28 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 MuiButton-disabled-21 c0"
  disabled={true}
  onBlur={[Function]}
  onClick={[MockFunction]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="-1"
  type="button"
>
  <span
    className="MuiButton-label-2"
  >
    Disabled
  </span>
</button>
`;

exports[`renders outline variant 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: black;
  padding: 0.5rem;
  border: solid 0.125rem currentColor;
  border-radius: 0.125rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<a
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 c0"
  href="#"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
>
  <span
    className="MuiButton-label-2"
  >
    Link
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</a>
`;

exports[`renders submit type 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: black;
  padding: 0.5rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<button
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 c0"
  disabled={false}
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex="0"
  type="submit"
>
  <span
    className="MuiButton-label-2"
  >
    Submit
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</button>
`;

exports[`renders wide 1`] = `
.c0.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: black;
  padding: 0.5rem;
  min-width: 13.75rem;
}

.c0.c0 svg {
  margin-right: 0.5rem;
}

<a
  className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6 c0"
  href="#"
  onBlur={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  role="button"
  tabIndex="0"
>
  <span
    className="MuiButton-label-2"
  >
    Link
  </span>
  <span
    className="MuiTouchRipple-root-30"
  />
</a>
`;
